File: stacks/node/base/tests/integration-tests.tpl.js
/**
 * Template: integration-tests.tpl.js
 * Purpose: integration-tests template
 * Stack: node
 * Tier: base
 */

# Universal Template System - Node Stack
# Generated: 2025-12-10
# Purpose: Testing utilities
# Tier: base
# Stack: node
# Category: testing

// -----------------------------------------------------------------------------
// FILE: integration-tests.tpl.js
// PURPOSE: Integration testing patterns for Node.js projects
// USAGE: Test interactions between multiple components and services
// DEPENDENCIES: jest, supertest, nock, mongodb-memory-server
// AUTHOR: [[.Author]]
// VERSION: [[.Version]]
// SINCE: [[.Version]]
// -----------------------------------------------------------------------------

/**
 * Node.js Integration Tests Template
 * Purpose: Integration testing patterns for Node.js projects
 * Usage: Test interactions between multiple components and services
 */

const request = require('supertest');
const nock = require('nock');
const { MongoMemoryServer } = require('mongodb-memory-server');
const mongoose = require('mongoose');
const app = require('../app');
const User = require('../models/User');
const AuthService = require('../services/AuthService');
const UserService = require('../services/UserService');

describe('Integration Tests - API Endpoints', () => {
  let mongoServer;
  let authToken;

  beforeAll(async () => {
    // Start in-memory MongoDB
    mongoServer = await MongoMemoryServer.create();
    const mongoUri = mongoServer.getUri();
    await mongoose.connect(mongoUri);
  });

  afterAll(async () => {
    await mongoose.disconnect();
    await mongoServer.stop();
  });

  beforeEach(async () => {
    // Clear database before each test
    await User.deleteMany({});
  });

  describe('User Authentication Flow', () => {
    test('should complete user registration and login flow', async () => {
      // Arrange
      const userData = {
        email: 'test@example.com',
        password: 'password123',
        name: 'Test User'
      };

      // Act - Register user
      const registerResponse = await request(app)
        .post('/api/auth/register')
        .send(userData)
        .expect(201);

      // Assert - Registration successful
      expect(registerResponse.body.email).toBe(userData.email);
      expect(registerResponse.body.id).toBeDefined();
      expect(registerResponse.body.password).toBeUndefined(); // Password should not be returned

      // Act - Login with registered user
      const loginResponse = await request(app)
        .post('/api/auth/login')
        .send({
          email: userData.email,
          password: userData.password
        })
        .expect(200);

      // Assert - Login successful
      expect(loginResponse.body.token).toBeDefined();
      expect(loginResponse.body.user.email).toBe(userData.email);
      authToken = loginResponse.body.token;
    });

    test('should protect routes with authentication middleware', async () => {
      // Act - Try to access protected route without token
      const response = await request(app)
        .get('/api/users/profile')
        .expect(401);

      // Assert
      expect(response.body.error).toContain('unauthorized');
    });

    test('should allow access to protected routes with valid token', async () => {
      // Arrange - Create and authenticate user
      const userData = {
        email: 'protected@example.com',
        password: 'password123',
        name: 'Protected User'
      };

      await request(app)
        .post('/api/auth/register')
        .send(userData);

      const loginResponse = await request(app)
        .post('/api/auth/login')
        .send({
          email: userData.email,
          password: userData.password
        });

      const token = loginResponse.body.token;

      // Act - Access protected route with token
      const response = await request(app)
        .get('/api/users/profile')
        .set('Authorization', `Bearer ${token}`)
        .expect(200);

      // Assert
      expect(response.body.email).toBe(userData.email);
    });
  });

  describe('User Management Integration', () => {
    let userToken;

    beforeEach(async () => {
      // Create and authenticate a user for each test
      const userData = {
        email: 'user@example.com',
        password: 'password123',
        name: 'Regular User'
      };

      await request(app)
        .post('/api/auth/register')
        .send(userData);

      const loginResponse = await request(app)
        .post('/api/auth/login')
        .send({
          email: userData.email,
          password: userData.password
        });

      userToken = loginResponse.body.token;
    });

    test('should update user profile', async () => {
      // Arrange
      const updateData = {
        name: 'Updated Name',
        bio: 'Updated bio'
      };

      // Act
      const response = await request(app)
        .put('/api/users/profile')
        .set('Authorization', `Bearer ${userToken}`)
        .send(updateData)
        .expect(200);

      // Assert
      expect(response.body.name).toBe(updateData.name);
      expect(response.body.bio).toBe(updateData.bio);
    });

    test('should handle file upload for user avatar', async () => {
      // Act
      const response = await request(app)
        .post('/api/users/avatar')
        .set('Authorization', `Bearer ${userToken}`)
        .attach('avatar', 'tests/fixtures/test-avatar.jpg')
        .expect(200);

      // Assert
      expect(response.body.avatarUrl).toBeDefined();
      expect(response.body.avatarUrl).toContain('.jpg');
    });

    test('should delete user account', async () => {
      // Act
      await request(app)
        .delete('/api/users/account')
        .set('Authorization', `Bearer ${userToken}`)
        .expect(200);

      // Assert - User should no longer be able to login
      await request(app)
        .post('/api/auth/login')
        .send({
          email: 'user@example.com',
          password: 'password123'
        })
        .expect(401);
    });
  });

  describe('Database Transactions', () => {
    test('should handle complex transactions correctly', async () => {
      // Arrange
      const userService = new UserService();
      const transactionData = {
        user: {
          email: 'transaction@example.com',
          password: 'password123',
          name: 'Transaction User'
        },
        profile: {
          bio: 'User bio',
          preferences: { theme: 'dark', language: 'en' }
        },
        settings: {
          notifications: true,
          privacy: { public: false }
        }
      };

      // Act
      const result = await userService.createUserWithProfile(transactionData);

      // Assert
      expect(result.user.id).toBeDefined();
      expect(result.profile.userId).toBe(result.user.id);
      expect(result.settings.userId).toBe(result.user.id);

      // Verify data consistency
      const user = await User.findById(result.user.id);
      expect(user).toBeTruthy();
    });

    test('should rollback on transaction failure', async () => {
      // Arrange
      const userService = new UserService();
      const invalidData = {
        user: {
          email: 'invalid@example.com',
          password: 'password123',
          name: 'Invalid User'
        },
        profile: null, // This should cause transaction to fail
        settings: {
          notifications: true
        }
      };

      // Act & Assert
      await expect(userService.createUserWithProfile(invalidData))
        .rejects.toThrow();

      // Verify no data was created
      const user = await User.findOne({ email: 'invalid@example.com' });
      expect(user).toBeNull();
    });
  });
});

describe('Integration Tests - External Services', () => {
  describe('Email Service Integration', () => {
    test('should send welcome email after registration', async () => {
      // Arrange - Mock email service
      const emailServiceMock = {
        sendWelcomeEmail: jest.fn().mockResolvedValue({ messageId: 'msg-123' })
      };

      // Mock the email service module
      jest.doMock('../services/EmailService', () => emailServiceMock);

      const userData = {
        email: 'welcome@example.com',
        password: 'password123',
        name: 'Welcome User'
      };

      // Act
      await request(app)
        .post('/api/auth/register')
        .send(userData);

      // Assert
      expect(emailServiceMock.sendWelcomeEmail).toHaveBeenCalledWith(userData.email, userData.name);
    });

    test('should handle email service failures gracefully', async () => {
      // Arrange - Mock email service failure
      const emailServiceMock = {
        sendWelcomeEmail: jest.fn().mockRejectedValue(new Error('Email service down'))
      };

      jest.doMock('../services/EmailService', () => emailServiceMock);

      const userData = {
        email: 'fail@example.com',
        password: 'password123',
        name: 'Fail User'
      };

      // Act
      const response = await request(app)
        .post('/api/auth/register')
        .send(userData)
        .expect(201);

      // Assert - User should still be created even if email fails
      expect(response.body.email).toBe(userData.email);
    });
  });

  describe('Payment Service Integration', () => {
    test('should process payment with external provider', async () => {
      // Arrange - Mock payment provider
      nock('https://api.stripe.com')
        .post('/v1/charges')
        .reply(200, {
          id: 'ch_123',
          status: 'succeeded',
          amount: 2000
        });

      const paymentData = {
        amount: 2000,
        currency: 'usd',
        source: 'tok_visa'
      };

      // Act
      const response = await request(app)
        .post('/api/payments/charge')
        .send(paymentData)
        .expect(200);

      // Assert
      expect(response.body.status).toBe('succeeded');
      expect(response.body.chargeId).toBe('ch_123');
    });

    test('should handle payment failures', async () => {
      // Arrange - Mock payment failure
      nock('https://api.stripe.com')
        .post('/v1/charges')
        .reply(400, {
          error: {
            type: 'card_error',
            message: 'Your card was declined.'
          }
        });

      const paymentData = {
        amount: 2000,
        currency: 'usd',
        source: 'tok_declined'
      };

      // Act
      const response = await request(app)
        .post('/api/payments/charge')
        .send(paymentData)
        .expect(400);

      // Assert
      expect(response.body.error).toContain('declined');
    });
  });
});

describe('Integration Tests - Caching', () => {
  let redisClient;

  beforeAll(async () => {
    // Setup Redis client for testing
    redisClient = require('../config/redis');
  });

  test('should cache and retrieve user data', async () => {
    // Arrange
    const userData = {
      email: 'cache@example.com',
      password: 'password123',
      name: 'Cache User'
    };

    // Act - Create user
    const createResponse = await request(app)
      .post('/api/auth/register')
      .send(userData);

    const userId = createResponse.body.id;

    // Get user data (should cache it)
    const getResponse = await request(app)
      .get(`/api/users/${userId}`)
      .expect(200);

    // Get user data again (should come from cache)
    const cachedResponse = await request(app)
      .get(`/api/users/${userId}`)
      .expect(200);

    // Assert
    expect(getResponse.body.email).toBe(userData.email);
    expect(cachedResponse.body.email).toBe(userData.email);
    
    // Verify cache was used (check for cache headers or timing)
    expect(cachedResponse.headers['x-cache']).toBe('hit');
  });

  test('should invalidate cache on user update', async () => {
    // Arrange
    const userData = {
      email: 'invalidate@example.com',
      password: 'password123',
      name: 'Original Name'
    };

    // Create user and get token
    const createResponse = await request(app)
      .post('/api/auth/register')
      .send(userData);

    const loginResponse = await request(app)
      .post('/api/auth/login')
      .send({
        email: userData.email,
        password: userData.password
      });

    const token = loginResponse.body.token;
    const userId = createResponse.body.id;

    // Get user data (caches it)
    await request(app)
      .get(`/api/users/${userId}`)
      .expect(200);

    // Update user
    await request(app)
      .put('/api/users/profile')
      .set('Authorization', `Bearer ${token}`)
      .send({ name: 'Updated Name' })
      .expect(200);

    // Get user data again (should not use stale cache)
    const updatedResponse = await request(app)
      .get(`/api/users/${userId}`)
      .expect(200);

    // Assert
    expect(updatedResponse.body.name).toBe('Updated Name');
    expect(updatedResponse.headers['x-cache']).toBe('miss');
  });
});

describe('Integration Tests - File Storage', () => {
  test('should upload and serve files from cloud storage', async () => {
    // Arrange
    const fileContent = Buffer.from('test file content');
    
    // Mock S3 upload
    nock('https://bucket.s3.amazonaws.com')
      .put('/uploads/test.txt')
      .reply(200, {
        Location: 'https://bucket.s3.amazonaws.com/uploads/test.txt',
        ETag: '"etag-123"'
      });

    // Act - Upload file
    const uploadResponse = await request(app)
      .post('/api/files/upload')
      .attach('file', fileContent, 'test.txt')
      .expect(200);

    // Assert
    expect(uploadResponse.body.url).toContain('test.txt');
    expect(uploadResponse.body.etag).toBe('"etag-123"');

    // Mock S3 download
    nock('https://bucket.s3.amazonaws.com')
      .get('/uploads/test.txt')
      .reply(200, fileContent);

    // Act - Download file
    const downloadResponse = await request(app)
      .get('/api/files/download/test.txt')
      .expect(200);

    // Assert
    expect(downloadResponse.body).toEqual(fileContent);
  });
});

describe('Integration Tests - WebSocket', () => {
  let server;
  let client;

  beforeAll(async () => {
    server = require('../server');
  });

  afterAll(async () => {
    if (server) {
      server.close();
    }
  });

  test('should handle real-time chat messages', async () => {
    // Arrange
    const WebSocket = require('ws');
    client = new WebSocket('ws://localhost:3000/chat');

    await new Promise((resolve) => {
      client.on('open', resolve);
    });

    // Act
    const message = {
      type: 'message',
      content: 'Hello, World!',
      userId: 'user123'
    };

    client.send(JSON.stringify(message));

    // Assert
    const response = await new Promise((resolve) => {
      client.on('message', (data) => {
        resolve(JSON.parse(data));
      });
    });

    expect(response.type).toBe('message');
    expect(response.content).toBe(message.content);
    expect(response.timestamp).toBeDefined();
  });

  test('should handle user join/leave events', async () => {
    // Arrange
    const WebSocket = require('ws');
    client = new WebSocket('ws://localhost:3000/chat');

    await new Promise((resolve) => {
      client.on('open', resolve);
    });

    // Act
    client.send(JSON.stringify({
      type: 'join',
      userId: 'user456',
      username: 'TestUser'
    }));

    // Assert
    const response = await new Promise((resolve) => {
      client.on('message', (data) => {
        const parsed = JSON.parse(data);
        if (parsed.type === 'user_joined') {
          resolve(parsed);
        }
      });
    });

    expect(response.type).toBe('user_joined');
    expect(response.userId).toBe('user456');
    expect(response.username).toBe('TestUser');
  });
});

describe('Integration Tests - Performance', () => {
  test('should handle concurrent requests efficiently', async () => {
    // Arrange
    const concurrentRequests = 50;
    const requests = [];

    // Act - Make concurrent requests
    for (let i = 0; i < concurrentRequests; i++) {
      requests.push(
        request(app)
          .get('/api/health')
          .expect(200)
      );
    }

    const startTime = Date.now();
    await Promise.all(requests);
    const endTime = Date.now();

    // Assert
    const totalTime = endTime - startTime;
    expect(totalTime).toBeLessThan(5000); // Should complete within 5 seconds
  });

  test('should handle large file uploads', async () => {
    // Arrange
    const largeFile = Buffer.alloc(10 * 1024 * 1024); // 10MB file

    // Act
    const startTime = Date.now();
    const response = await request(app)
      .post('/api/files/upload')
      .attach('file', largeFile, 'large-file.bin')
      .expect(200);
    const endTime = Date.now();

    // Assert
    expect(response.body.size).toBe(10 * 1024 * 1024);
    expect(endTime - startTime).toBeLessThan(10000); // Should upload within 10 seconds
  });
});

describe('Integration Tests - Error Handling', () => {
  test('should handle database connection failures', async () => {
    // Arrange - Disconnect database
    await mongoose.connection.close();

    // Act
    const response = await request(app)
      .get('/api/users/profile')
      .expect(500);

    // Assert
    expect(response.body.error).toContain('database');
  });

  test('should handle external API timeouts', async () => {
    // Arrange - Mock slow external API
    nock('https://api.external.com')
      .get('/data')
      .delayConnection(5000) // 5 second delay
      .reply(200, { data: 'response' });

    // Act
    const response = await request(app)
      .get('/api/external/data')
      .expect(408); // Request timeout

    // Assert
    expect(response.body.error).toContain('timeout');
  });
});

// Mock implementations for testing
class MockEmailService {
  async sendWelcomeEmail(email, name) {
    return { messageId: 'msg-123' };
  }
}

class MockPaymentService {
  async charge(amount, currency, source) {
    return {
      id: 'ch_123',
      status: 'succeeded',
      amount
    };
  }
}

module.exports = {
  MockEmailService,
  MockPaymentService
};

File: stacks/python/base/tests/integration-tests.tpl.py
# Universal Template System - Python Stack
# Generated: 2025-12-10
# Purpose: Testing utilities
# Tier: base
# Stack: python
# Category: testing

#!/usr/bin/env python3
# -----------------------------------------------------------------------------
# FILE: integration-tests.tpl.py
# PURPOSE: Integration testing patterns for Python projects
# USAGE: Test interactions between multiple components and services
# DEPENDENCIES: pytest, pytest-asyncio, httpx, testcontainers
# AUTHOR: [[.Author]]
# VERSION: [[.Version]]
# SINCE: [[.Version]]
# -----------------------------------------------------------------------------

"""
Python Integration Tests Template
Purpose: Integration testing patterns for Python projects
Usage: Test interactions between multiple components and services
"""

import pytest
import asyncio
import httpx
from unittest.mock import patch, AsyncMock
from testcontainers.postgres import PostgresContainer
from testcontainers.redis import RedisContainer
from fastapi.testclient import TestClient
from sqlalchemy import create_engine
from sqlalchemy.orm import sessionmaker

# Import your application modules here
# from your_app.main import app
# from your_app.database import get_db, Base
# from your_app.services.auth_service import AuthService
# from your_app.services.user_service import UserService
# from your_app.repositories.user_repository import UserRepository

@pytest.fixture(scope="session")
def event_loop():
    """Create an instance of the default event loop for the test session."""
    loop = asyncio.get_event_loop_policy().new_event_loop()
    yield loop
    loop.close()

@pytest.fixture(scope="function")
async def postgres_container():
    """Start PostgreSQL container for integration tests."""
    with PostgresContainer("postgres:13") as postgres:
        yield postgres

@pytest.fixture(scope="function")
async def redis_container():
    """Start Redis container for integration tests."""
    with RedisContainer("redis:6") as redis:
        yield redis

@pytest.fixture(scope="function")
async def test_db(postgres_container):
    """Create test database session."""
    engine = create_engine(postgres_container.get_connection_url())
    TestingSessionLocal = sessionmaker(autocommit=False, autoflush=False, bind=engine)
    
    # Create tables
    Base.metadata.create_all(bind=engine)
    
    db = TestingSessionLocal()
    try:
        yield db
    finally:
        db.close()

@pytest.fixture(scope="function")
async def test_client(test_db):
    """Create test client with database dependency override."""
    def override_get_db():
        try:
            yield test_db
        finally:
            test_db.close()
    
    app.dependency_overrides[get_db] = override_get_db
    with TestClient(app) as client:
        yield client
    app.dependency_overrides.clear()

class TestUserAuthenticationIntegration:
    """Integration tests for user authentication workflow."""
    
    async def test_complete_user_registration_flow(self, test_client: TestClient):
        """Test complete user registration from API to database."""
        # Arrange
        user_data = {
            "email": "test@example.com",
            "password": "password123",
            "name": "Test User"
        }
        
        # Act - Register user
        response = test_client.post("/api/auth/register", json=user_data)
        
        # Assert - Registration successful
        assert response.status_code == 201
        assert response.json()["email"] == user_data["email"]
        assert "id" in response.json()
        
        # Act - Login with registered user
        login_data = {
            "email": user_data["email"],
            "password": user_data["password"]
        }
        login_response = test_client.post("/api/auth/login", json=login_data)
        
        # Assert - Login successful
        assert login_response.status_code == 200
        assert "access_token" in login_response.json()
        assert login_response.json()["token_type"] == "bearer"
        
        # Act - Access protected endpoint
        token = login_response.json()["access_token"]
        headers = {"Authorization": f"Bearer {token}"}
        profile_response = test_client.get("/api/users/profile", headers=headers)
        
        # Assert - Protected endpoint accessible
        assert profile_response.status_code == 200
        assert profile_response.json()["email"] == user_data["email"]

    async def test_user_session_management(self, test_client: TestClient):
        """Test user session creation and management."""
        # Arrange
        user_data = {
            "email": "session@example.com",
            "password": "password123",
            "name": "Session User"
        }
        
        # Register user
        test_client.post("/api/auth/register", json=user_data)
        
        # Act - Login and create session
        login_response = test_client.post("/api/auth/login", json={
            "email": user_data["email"],
            "password": user_data["password"]
        })
        
        # Assert - Session created
        assert login_response.status_code == 200
        token = login_response.json()["access_token"]
        
        # Act - Verify session in database
        headers = {"Authorization": f"Bearer {token}"}
        session_response = test_client.get("/api/auth/sessions", headers=headers)
        
        # Assert - Session exists
        assert session_response.status_code == 200
        sessions = session_response.json()
        assert len(sessions) > 0
        assert sessions[0]["user_id"] is not None

    async def test_password_reset_flow(self, test_client: TestClient):
        """Test complete password reset workflow."""
        # Arrange
        user_data = {
            "email": "reset@example.com",
            "password": "oldpassword",
            "name": "Reset User"
        }
        
        # Register user
        test_client.post("/api/auth/register", json=user_data)
        
        # Act - Request password reset
        reset_request = test_client.post("/api/auth/reset-password", json={
            "email": user_data["email"]
        })
        
        # Assert - Reset token generated
        assert reset_request.status_code == 200
        reset_token = reset_request.json()["reset_token"]
        
        # Act - Reset password
        reset_confirm = test_client.post("/api/auth/reset-password/confirm", json={
            "token": reset_token,
            "new_password": "newpassword123"
        })
        
        # Assert - Password reset successful
        assert reset_confirm.status_code == 200
        
        # Act - Login with new password
        login_response = test_client.post("/api/auth/login", json={
            "email": user_data["email"],
            "password": "newpassword123"
        })
        
        # Assert - Login successful with new password
        assert login_response.status_code == 200

class TestDatabaseIntegration:
    """Integration tests for database operations."""
    
    async def test_user_crud_operations(self, test_db):
        """Test CRUD operations with real database."""
        # Arrange
        user_repo = UserRepository(test_db)
        user_data = {
            "email": "crud@example.com",
            "password": "password123",
            "name": "CRUD User"
        }
        
        # Act - Create user
        created_user = await user_repo.create(user_data)
        
        # Assert - User created
        assert created_user.id is not None
        assert created_user.email == user_data["email"]
        
        # Act - Read user
        retrieved_user = await user_repo.get_by_id(created_user.id)
        
        # Assert - User retrieved
        assert retrieved_user is not None
        assert retrieved_user.email == user_data["email"]
        
        # Act - Update user
        update_data = {"name": "Updated User"}
        updated_user = await user_repo.update(created_user.id, update_data)
        
        # Assert - User updated
        assert updated_user.name == update_data["name"]
        
        # Act - Delete user
        deleted = await user_repo.delete(created_user.id)
        
        # Assert - User deleted
        assert deleted is True
        
        # Verify deletion
        deleted_user = await user_repo.get_by_id(created_user.id)
        assert deleted_user is None

    async def test_database_transactions(self, test_db):
        """Test database transaction handling."""
        # Arrange
        user_repo = UserRepository(test_db)
        
        # Act - Test successful transaction
        async with test_db.begin():
            user1 = await user_repo.create({
                "email": "trans1@example.com",
                "password": "password123",
                "name": "Transaction User 1"
            })
            user2 = await user_repo.create({
                "email": "trans2@example.com",
                "password": "password123",
                "name": "Transaction User 2"
            })
        
        # Assert - Both users created
        assert user1.id is not None
        assert user2.id is not None
        
        # Act - Test failed transaction
        try:
            async with test_db.begin():
                user3 = await user_repo.create({
                    "email": "trans3@example.com",
                    "password": "password123",
                    "name": "Transaction User 3"
                })
                # Force an error
                raise ValueError("Transaction failed")
        except ValueError:
            pass
        
        # Assert - User3 not created due to transaction rollback
        user3_retrieved = await user_repo.get_by_email("trans3@example.com")
        assert user3_retrieved is None

class TestAPIIntegration:
    """Integration tests for API endpoints."""
    
    async def test_cors_headers(self, test_client: TestClient):
        """Test CORS headers are properly set."""
        # Act
        response = test_client.options("/api/users/profile")
        
        # Assert
        assert "access-control-allow-origin" in response.headers
        assert "access-control-allow-methods" in response.headers

    async def test_rate_limiting(self, test_client: TestClient):
        """Test API rate limiting functionality."""
        # Act - Make multiple requests quickly
        responses = []
        for _ in range(10):
            response = test_client.post("/api/auth/login", json={
                "email": "test@example.com",
                "password": "wrongpassword"
            })
            responses.append(response)
        
        # Assert - Should be rate limited after certain attempts
        rate_limited_responses = [r for r in responses if r.status_code == 429]
        assert len(rate_limited_responses) > 0

    async def test_api_versioning(self, test_client: TestClient):
        """Test API versioning works correctly."""
        # Act - Test different API versions
        v1_response = test_client.get("/api/v1/users/profile")
        v2_response = test_client.get("/api/v2/users/profile")
        
        # Assert - Different versions should respond appropriately
        assert v1_response.status_code in [200, 401]  # May require auth
        assert v2_response.status_code in [200, 401]  # May require auth

class TestExternalServiceIntegration:
    """Integration tests for external service integrations."""
    
    @patch('httpx.AsyncClient.get')
    async def test_third_party_api_integration(self, mock_get):
        """Test integration with third-party APIs."""
        # Arrange
        mock_response = httpx.Response(
            status_code=200,
            json={"data": "external_data", "status": "success"}
        )
        mock_get.return_value = mock_response
        
        # Act
        external_service = ExternalService()
        result = await external_service.fetch_data()
        
        # Assert
        assert result["status"] == "success"
        assert result["data"] == "external_data"
        mock_get.assert_called_once()

    async def test_email_service_integration(self):
        """Test email service integration."""
        # Arrange
        email_service = EmailService()
        
        # Act - Send test email
        result = await email_service.send_email(
            to="test@example.com",
            subject="Test Email",
            body="This is a test email"
        )
        
        # Assert
        assert result["status"] == "sent"
        assert "message_id" in result

class TestCacheIntegration:
    """Integration tests for caching functionality."""
    
    async def test_redis_caching(self, redis_container):
        """Test Redis caching integration."""
        # Arrange
        cache_service = RedisCacheService(redis_container.get_connection_url())
        
        # Act - Cache data
        await cache_service.set("test_key", {"data": "test_value"}, ttl=60)
        
        # Retrieve cached data
        cached_data = await cache_service.get("test_key")
        
        # Assert
        assert cached_data is not None
        assert cached_data["data"] == "test_value"
        
        # Act - Delete cached data
        await cache_service.delete("test_key")
        deleted_data = await cache_service.get("test_key")
        
        # Assert
        assert deleted_data is None

    async def test_cache_invalidation(self, redis_container):
        """Test cache invalidation strategies."""
        # Arrange
        cache_service = RedisCacheService(redis_container.get_connection_url())
        
        # Cache multiple related keys
        await cache_service.set("user:1:profile", {"name": "User 1"})
        await cache_service.set("user:1:settings", {"theme": "dark"})
        await cache_service.set("user:1:posts", [{"id": 1, "title": "Post 1"}])
        
        # Act - Invalidate all user-related cache
        await cache_service.invalidate_pattern("user:1:*")
        
        # Assert - All user-related cache cleared
        profile = await cache_service.get("user:1:profile")
        settings = await cache_service.get("user:1:settings")
        posts = await cache_service.get("user:1:posts")
        
        assert profile is None
        assert settings is None
        assert posts is None

class TestBackgroundTaskIntegration:
    """Integration tests for background task processing."""
    
    async def test_async_task_processing(self):
        """Test async background task processing."""
        # Arrange
        task_queue = TaskQueue()
        
        # Act - Enqueue tasks
        task1_id = await task_queue.enqueue("process_data", {"data": "test1"})
        task2_id = await task_queue.enqueue("process_data", {"data": "test2"})
        
        # Process tasks
        results = await task_queue.process_batch(max_tasks=2)
        
        # Assert
        assert len(results) == 2
        assert all(result["status"] == "completed" for result in results)

    async def test_task_failure_handling(self):
        """Test background task failure handling."""
        # Arrange
        task_queue = TaskQueue()
        
        # Act - Enqueue failing task
        task_id = await task_queue.enqueue("failing_task", {"should_fail": True})
        
        # Process task
        result = await task_queue.process_task(task_id)
        
        # Assert
        assert result["status"] == "failed"
        assert "error" in result

class TestFileStorageIntegration:
    """Integration tests for file storage operations."""
    
    async def test_file_upload_workflow(self):
        """Test complete file upload workflow."""
        # Arrange
        storage_service = S3StorageService()
        file_content = b"test file content"
        
        # Act - Upload file
        upload_result = await storage_service.upload(
            file_content=file_content,
            filename="test.txt",
            content_type="text/plain"
        )
        
        # Assert
        assert "url" in upload_result
        assert upload_result["size"] == len(file_content)
        
        # Act - Download file
        downloaded_content = await storage_service.download(upload_result["key"])
        
        # Assert
        assert downloaded_content == file_content
        
        # Act - Delete file
        delete_result = await storage_service.delete(upload_result["key"])
        
        # Assert
        assert delete_result is True

# Example service classes for testing
class ExternalService:
    async def fetch_data(self):
        async with httpx.AsyncClient() as client:
            response = await client.get("https://api.example.com/data")
            return response.json()

class EmailService:
    async def send_email(self, to: str, subject: str, body: str):
        # Mock email service implementation
        return {
            "status": "sent",
            "message_id": "msg_12345",
            "to": to,
            "subject": subject
        }

class RedisCacheService:
    def __init__(self, redis_url: str):
        self.redis_url = redis_url
    
    async def set(self, key: str, value: dict, ttl: int = 3600):
        # Mock Redis implementation
        pass
    
    async def get(self, key: str):
        # Mock Redis implementation
        return None
    
    async def delete(self, key: str):
        # Mock Redis implementation
        pass
    
    async def invalidate_pattern(self, pattern: str):
        # Mock Redis implementation
        pass

class TaskQueue:
    async def enqueue(self, task_name: str, task_data: dict):
        # Mock task queue implementation
        return f"task_{task_name}_{hash(str(task_data))}"
    
    async def process_batch(self, max_tasks: int = 10):
        # Mock batch processing
        return [{"status": "completed", "task_id": f"task_{i}"} for i in range(max_tasks)]
    
    async def process_task(self, task_id: str):
        # Mock single task processing
        return {"status": "completed", "task_id": task_id}

class S3StorageService:
    async def upload(self, file_content: bytes, filename: str, content_type: str):
        # Mock S3 upload
        return {
            "key": f"uploads/{filename}",
            "url": f"https://bucket.s3.amazonaws.com/uploads/{filename}",
            "size": len(file_content)
        }
    
    async def download(self, key: str):
        # Mock S3 download
        return b"downloaded content"
    
    async def delete(self, key: str):
        # Mock S3 delete
        return True

if __name__ == "__main__":
    pytest.main([
        __file__, 
        "-v", 
        "--cov=your_app", 
        "--cov-report=html",
        "--cov-report=term-missing"
    ])
